### PR TITLE
Feature/agent client mapping

### PR DESCRIPTION
**Pull request by RobinAtherton**
Added Container-Structure for saving Clientgroups.
Added accepting and rejecting for AgentHandshakeHandler.

### CODE REVIEW

File: src/main/java/org/varroa/load/MqttClientPool.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.mqttbee.mqtt.MqttClient;
+
+import java.util.LinkedList;
+
+/**
+ * @author Silvio Giebl
+ * @author Robin Atherton
+ */
+public class MqttClientPool {
```
Comment (YannickWeber) [inline src/main/java/org/varroa/load/MqttClientPool.java]: Please add Java Doc and describe the function of this class.

File: src/main/java/org/varroa/load/MqttClientPool.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        final MqttClientSubPool lastSubPool = groupSubPools[groupSubPools.length - 1];
+        final int lastSize = mqttClientGroup.count % MqttClientSubPool.SIZE;
+        if (lastSize == 0) {
+            subPools.addLast(lastSubPool);
+        } else {
+            moveToMosaic(lastSubPool, 0, lastSize);
+        }
+    }
+
+    public @NotNull MqttClientGroup take(final int count) {
```
Comment (YannickWeber) [inline src/main/java/org/varroa/load/MqttClientPool.java]: javadoc please

File: src/main/java/org/varroa/load/MqttClientPool.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+            if (count < mosaicSize) {
+                moveToMosaic(lastSubPool, count, mosaicSize);
+            }
+        }
+
+        return new MqttClientGroup(groupSubPools, count);
+    }
+
+    private void moveToMosaic(final @NotNull MqttClientSubPool subPool, final int from, final int to) {
```
Comment (YannickWeber) [inline src/main/java/org/varroa/load/MqttClientPool.java]: add java doc

File: src/main/java/org/varroa/load/MqttClientSubPool.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package org.varroa.load;
+
+import org.jetbrains.annotations.NotNull;
+import org.mqttbee.mqtt.MqttClient;
+
+/**
+ * @author Silvio Giebl
+ * @author Robin Atherton
+ */
+class MqttClientSubPool {
```
Comment (YannickWeber) [inline src/main/java/org/varroa/load/MqttClientSubPool.java]: java doc please

File: src/main/java/org/varroa/load/MqttClientSubPool.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+    static final int SIZE = 16;
```
Comment (YannickWeber) [inline src/main/java/org/varroa/load/MqttClientSubPool.java]: why 16?

File: src/main/java/org/varroa/remote/agent/AgentHandshakeHandler.java [hunk 6]
```diff
@@ -17,19 +27,106 @@
...
+
+        final boolean firstIsConnect = commands.get(0) instanceof ConnectCommand;
+        if (firstIsConnect) {
+            accepted = true;
+        } else {
+            final MqttClientPool mqttClientPool = mqttClientPoolMap.get(chunk.getClientGroupId());
+            if ((mqttClientPool != null) && (mqttClientPool.size() >= chunk.getCount())) {
+                accepted = true;
+            }
+            // TODO reserve client count
```
Comment (YannickWeber) [inline src/main/java/org/varroa/remote/agent/AgentHandshakeHandler.java]: ?

File: src/test/java/org/varroa/load/MqttClientGroupTest.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ * @author Silvio Giebl
+ * @author Robin Atherton
+ */
+class MqttClientGroupTest {
+
+    @Test
+    void countSmallerThanSubPoolSize() {
+        final MqttClientGroup group = new MqttClientGroup(10);
+        assertEquals(10, group.getCount());
+        assertEquals(1, group.clientSubPools.length);
```
Comment (YannickWeber) [inline src/test/java/org/varroa/load/MqttClientGroupTest.java]: add messages

File: src/main/java/org/varroa/remote/agent/AgentHandshakeHandler.java [hunk 8]
```diff
@@ -17,19 +27,106 @@
...
+            reject(ctx, msg.getId());
+        }
+    }
+
+    private void accept(final @NotNull ChannelHandlerContext ctx, final int id) {
+        ctx.writeAndFlush(new ChunkAcceptedMessage(id));
+    }
+
+    private void reject(final @NotNull ChannelHandlerContext ctx, final int id) {
+        ctx.writeAndFlush(new ChunkAcceptedMessage(id));
```
Comment (s-gheldd) [inline src/main/java/org/varroa/remote/agent/AgentHandshakeHandler.java]: ChunkRejectedMessage

File: src/main/java/org/varroa/remote/agent/AgentHandshakeHandler.java [hunk 9]
```diff
@@ -17,19 +27,106 @@
...
+                mqttClientPool = new MqttClientPool();
+                mqttClientPoolMap.put(chunk.getClientGroupId(), mqttClientPool);
+            }
+            mqttClientPool.offer(finishedEvent.mqttClientGroup);
+        }
+
+        ctx.writeAndFlush(new ChunkFinishedMessage(finishedEvent.chunkMessage.getId()));
+    }
+
+    public static class FinishedEvent {
```
Comment (s-gheldd) [inline src/main/java/org/varroa/remote/agent/AgentHandshakeHandler.java]: @Immutable

File: src/main/java/org/varroa/remote/agent/AgentHandshakeHandler.java [hunk 10]
```diff
@@ -1,13 +1,23 @@
 package org.varroa.remote.agent;
 
+import com.google.common.collect.ImmutableList;
```
Comment (s-gheldd) [inline src/main/java/org/varroa/remote/agent/AgentHandshakeHandler.java]: Bitte noch Tests nachpflegen.

**Comment by s-gheldd** [comment]
Nice PR. Can be merged!

**Pull request merged**